pilopt/src/lib.rs

```diff
@@ -441,55 +437,59 @@ fn remove_constant_witness_columns<T: FieldElement>(pil_file: &mut Analyzed<T>)
         .filter(|&id| (id.kind == IdentityKind::Polynomial))
         .map(|id| id.expression_for_poly_id())
         .filter_map(constrained_to_constant)
-        .collect::<BTreeMap<PolyID, _>>();
+        .collect::<Vec<(String, PolyID, _)>>();
     // We cannot remove arrays or array elements, so filter them out.
     let columns = pil_file
         .committed_polys_in_source_order()
         .iter()
         .filter(|&(s, _)| (!s.is_array()))
         .map(|(s, _)| s.into())
         .collect::<HashSet<PolyID>>();
-    constant_polys.retain(|id, _| columns.contains(id));
+    constant_polys.retain(|(_, id, _)| columns.contains(id));
 
-    substitute_polynomial_references(pil_file, &constant_polys);
+    substitute_polynomial_references(pil_file, constant_polys);
 }
 
 /// Substitutes all references to certain polynomials by the given field elements.
 fn substitute_polynomial_references<T: FieldElement>(
     pil_file: &mut Analyzed<T>,
-    substitutions: &BTreeMap<PolyID, BigUint>,
+    substitutions: Vec<(String, PolyID, BigUint)>,
 ) {
+    let substitutions_by_id = substitutions
+        .iter()
+        .map(|(_, id, value)| (*id, value.clone()))
+        .collect::<HashMap<PolyID, _>>();
+    let substitutions_by_name = substitutions
+        .into_iter()
+        .map(|(name, _, value)| (name, value))
+        .collect::<HashMap<String, _>>();
     pil_file.post_visit_expressions_in_definitions_mut(&mut |e: &mut Expression| {
         if let Expression::Reference(
             _,
-            Reference::Poly(PolynomialReference {
-                name: _,
-                poly_id: Some(poly_id),
-                type_args: _,
-            }),
+            Reference::Poly(PolynomialReference { name, type_args: _ }),
         ) = e
         {
-            if let Some(value) = substitutions.get(poly_id) {
+            if let Some(value) = substitutions_by_name.get(name) {
                 *e = Number {
-                    value: value.clone(),
-                    type_: Some(Type::Fe),
+                    value: (*value).clone(),
+                    type_: Some(Type::Expr),
                 }
                 .into();
             }
         }
     });
     pil_file.post_visit_expressions_in_identities_mut(&mut |e: &mut AlgebraicExpression<_>| {
         if let AlgebraicExpression::Reference(AlgebraicReference { poly_id, .. }) = e {
-            if let Some(value) = substitutions.get(poly_id) {
-                *e = AlgebraicExpression::Number(T::checked_from(value.clone()).unwrap());
+            if let Some(value) = substitutions_by_id.get(poly_id) {
+                *e = AlgebraicExpression::Number(T::checked_from((*value).clone()).unwrap());
             }
         }
     });
 }
 
 fn constrained_to_constant<T: FieldElement>(
     expr: &AlgebraicExpression<T>,
-) -> Option<(PolyID, BigUint)> {
+) -> Option<(String, PolyID, BigUint)> {
```

Comment:
It was not clear to me that this is still a key and a value.
```suggestion
) -> Option<((String, PolyID), BigUint)> {
```
Or introduce a `SubstitutionKey` struct